src/main/java/at/nonblocking/maven/nonsnapshot/impl/UpstreamDependencyHandlerDefaultImpl.java

```diff
@@ -48,54 +50,131 @@
+                                            MavenPomHandler mavenPomHandler,
+                                            RepositorySystem repositorySystem,
+                                            RepositorySystemSession repositorySystemSession,
+                                            List<RemoteRepository> remoteRepositories)
+  {
+    try {
+      LOG.info("Resolving artifact: {}", upstreamDependencyString);
+      ArtifactRequest artifactRequest = new ArtifactRequest(new DefaultArtifact(upstreamDependencyString), remoteRepositories, null);
+      ArtifactResult result = repositorySystem.resolveArtifact(repositorySystemSession, artifactRequest);
+      LOG.info("Artifact {} resolved parsing its dependencies", upstreamDependencyString);
+      MavenModule bomMavenModule = mavenPomHandler.readArtifact(result.getArtifact().getFile());
+      for(MavenModuleDependency indirectDependency : bomMavenModule.getDependencies()) {
+        MavenArtifact indirectDependencyArtifact = indirectDependency.getArtifact();
+        String indirectDependencyScope = indirectDependency.getScope();
+        processedUpstreamDependencies.add(makeProcessedUpstreamDependencyFromMavenArtifact(indirectDependencyArtifact));
+        LOG.info(
+          "Indirect dependency: {} (scope: {}) found and added", indirectDependencyArtifact.toString(), indirectDependencyScope);
+        if(indirectDependencyArtifact.getType() != null && indirectDependencyScope != null &&
+           indirectDependencyArtifact.getType().equals("pom") && indirectDependencyScope.equals("import"))
+        {
+          LOG.info(
+            "Indirect dependency: {} is of pom type and has import scope - it will be treated as bom upstream dependency",
+            indirectDependencyArtifact.toString()
+          );
+          ProcessBomUpstreamDependency(
+            indirectDependencyArtifact.toString(), processedUpstreamDependencies, mavenPomHandler, repositorySystem, repositorySystemSession, remoteRepositories);
+        }
+
 
-    for (String upstreamDependencyString : upstreamDependencyStrings) {
-      if (upstreamDependencyString.trim().isEmpty()) {
-        throw new NonSnapshotPluginException("Illegal upstreamDependency: " + upstreamDependencyString);
       }
+    } catch (ArtifactResolutionException e) {
+      throw new NonSnapshotDependencyResolverException("Couldn't resolve bom dependency: " + upstreamDependencyString, e);
+    }
+  }
 
-      Pattern groupPattern;
-      Pattern artifactPattern = null;
-      Integer versionMajor = null;
-      Integer versionMinor = null;
-      Integer versionIncrement = null;
+  @Override
+  public List<ProcessedUpstreamDependency> processDependencyList(List upstreamDependencies,
+                                                                 MavenPomHandler mavenPomHandler,
+                                                                 RepositorySystem repositorySystem,
```

Comment:
You are passing those three Repository parameters as method arguments in many places but always together. I'd suggest creating a wrapper structure for all three of them and pass them as one parameter.

Reply:
I just add this to one of methods. In other it was used in this way, so I don't decide to change too much.